ssh-ecdsa.c

```diff
@@ -65,30 +68,22 @@ ssh_ecdsa_size(const struct sshkey *key)
 static void
 ssh_ecdsa_cleanup(struct sshkey *k)
 {
-	EC_KEY_free(k->ecdsa);
-	k->ecdsa = NULL;
+	EVP_PKEY_free(k->pkey);
+	k->pkey = NULL;
 }
 
 static int
 ssh_ecdsa_equal(const struct sshkey *a, const struct sshkey *b)
 {
-	const EC_GROUP *grp_a, *grp_b;
-	const EC_POINT *pub_a, *pub_b;
-
-	if (a->ecdsa == NULL || b->ecdsa == NULL)
-		return 0;
-	if ((grp_a = EC_KEY_get0_group(a->ecdsa)) == NULL ||
-	    (grp_b = EC_KEY_get0_group(b->ecdsa)) == NULL)
-		return 0;
-	if ((pub_a = EC_KEY_get0_public_key(a->ecdsa)) == NULL ||
-	    (pub_b = EC_KEY_get0_public_key(b->ecdsa)) == NULL)
-		return 0;
-	if (EC_GROUP_cmp(grp_a, grp_b, NULL) != 0)
-		return 0;
-	if (EC_POINT_cmp(grp_a, pub_a, pub_b, NULL) != 0)
-		return 0;
+#if (OPENSSL_VERSION_NUMBER >= 0x30000000L)
+	if (EVP_PKEY_eq(a->pkey, b->pkey) == 1)
+		return 1;
+#else
+	if (EVP_PKEY_cmp(a->pkey, b->pkey) == 1)
+		return 1;
+#endif
```

Comment:
So this is a general comment about this entire change.  I made it in 455 and this is convenienty a different PR so the comment is now lost, and I'll state it again here.  The EVP_PKEY interface in 3.0 is not that different from the 1.1.1 level of API supported in libressl.  This is an example of a needless change because 3.0 STILL HAS EVP_PKEY_cmp and can do this just as it did before.  

The result here is a mess of #ifdef soup promoting the new API instead of the old one, merely because 3.0 is claiming to mark some funcitons as deprecated and the new api is preferred for new code.  this is fine, and we could even have discussions later about ensure *some* of these API's were universally supported, but that will not happen if this goes in as #ifdef soup trying to make everything the newest 3.0 way.. 

As I mentioned, I'd love to see this in a form where we simply have this coverted to the 1.1.1 EVP_PKEY api first. and at that point, look at what the minimal differences are to get this to compile on 3.0, which i expect are absolutely not this.

Reply:
I don't see any problem with removing #ifdef in this particular function (and the same for EC). In general I'd say we need some compatibility layer.